examples/android/routeguide/app/src/main/java/io/grpc/routeguideexample/RouteGuideActivity.java

```diff
@@ -0,0 +1,370 @@
+    private Button mRouteChatButton;
+    private TextView mResultText;
+    private ManagedChannel mChannel;
+    private RouteGuideGrpc.RouteGuideBlockingStub mBlockingStub;
+    private RouteGuideGrpc.RouteGuideStub mAsyncStub;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_routeguide);
+        mHostEdit = (EditText) findViewById(R.id.host_edit_text);
+        mPortEdit = (EditText) findViewById(R.id.port_edit_text);
+        mGetFeatureButton = (Button) findViewById(R.id.get_feature_button);
+        mListFeaturesButton = (Button) findViewById(R.id.list_features_button);
+        mRecordRouteButton = (Button) findViewById(R.id.record_route_button);
+        mRouteChatButton = (Button) findViewById(R.id.route_chat_button);
+        mResultText = (TextView) findViewById(R.id.result_text);
+    }
+
+    public void getFeature(View view) {
+        disableButtons();
+        new GrpcTask(new GrpcRunnable() {
+            @Override
+            public void run() throws Exception {
+                doGetFeature(409146138, -746188906);
+            }
+        }).execute();
+    }
+
+    public void listFeatures(View view) {
+        disableButtons();
+        new GrpcTask(new GrpcRunnable() {
+            @Override
+            public void run() throws Exception {
+                doListFeatures(400000000, -750000000, 420000000, -730000000);
+            }
+        }).execute();
+    }
+
+    public void recordRoute(View view) {
+        disableButtons();
+        new GrpcTask(new GrpcRunnable() {
+            @Override
+            public void run() throws Exception {
+                List<Point> points = new ArrayList<Point>();
+                points.add(Point.newBuilder()
+                        .setLatitude(407838351).setLongitude(-746143763).build());
+                points.add(Point.newBuilder()
+                        .setLatitude(408122808).setLongitude(-743999179).build());
+                points.add(Point.newBuilder()
+                        .setLatitude(413628156).setLongitude(-749015468).build());
+                doRecordRoute(points, 5);
+            }
+        }).execute();
+    }
+
+    public void routeChat(View view) {
+        disableButtons();
+        new GrpcTask(new GrpcRunnable() {
+            @Override
+            public void run() throws Exception {
+                doRouteChat();
+            }
+        }).execute();
+    }
+
+    private void disableButtons() {
+        ((InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE))
+                .hideSoftInputFromWindow(mHostEdit.getWindowToken(), 0);
+        mGetFeatureButton.setEnabled(false);
+        mListFeaturesButton.setEnabled(false);
+        mRecordRouteButton.setEnabled(false);
+        mRouteChatButton.setEnabled(false);
+    }
+
+    private class GrpcTask extends AsyncTask<Void, Void, String> {
```

Comment:
I've moved the channel into GrpcTask and pass stubs in the GrpcRunnable.run so it's not needed to mutate RouteGuideActivity fields.
Also made all the GrpcRunnable.run implementations to correctly throw exceptions instead of just logging them.